app/assets/javascripts/discourse/app/components/translation.gjs

```
@@ -0,0 +1,193 @@
+        parts.push(text.slice(currentIndex, match.index));
+      }
+
+      // Add the placeholder element, but only if the placeholder string we found matches
+      // the uniqueId we generated earlier for that placeholder.
+      if (this._placeholderKeys.get(match[1]) === match[0]) {
+        parts.push(this._placeholderElements.get(match[1]));
+      }
+
+      currentIndex = match.index + match[0].length;
+    }
+
+    // Add remaining text if any
+    if (currentIndex < text.length) {
+      parts.push(text.slice(currentIndex));
+    }
+
+    return parts;
+  }
+
+  /**
+   * Creates a curried TranslationPlaceholder component for a specific placeholder.
+   * This allows the placeholder component to be passed to the template's named block
+   * with the placeholder name already bound.
+   *
+   * @param {String} placeholder - The name of the placeholder to create a component for
+   * @returns {Component} A curried TranslationPlaceholder component with the placeholder name bound
+   */
+  @action
+  placeholderElement(placeholder) {
+    return curryComponent(
+      TranslationPlaceholder,
+      { placeholder, markAsRendered: this.markAsRendered },
+      getOwner(this)
+    );
+  }
+
+  /**
+   * Marks a placeholder as having been rendered with content.
+   * Called by the TranslationPlaceholder component when it renders.
+   *
+   * @param {String} name - The name of the placeholder that has been rendered
+   */
+  @action
+  markAsRendered(name) {
+    this._renderedPlaceholders.push(name);
+  }
+
+  /**
+   * Checks for any placeholders that were expected but not provided in the template, then
+   * inserts a warning message where that placeholder was supposed to be.
+   */
+  @action
+  checkPlaceholders() {
+    for (const [name, element] of this._placeholderElements) {
+      if (!this._renderedPlaceholders.includes(name)) {
+        element.innerText = `[missing ${this._placeholderAppearance.get(
+          name
+        )} placeholder]`;
```

I think we should also throw an error in the testing environment just like the `i18n` function:

https://github.com/discourse/discourse/blob/976886ab7cccbfd48277e3b5c98c9103e869e080/app/assets/javascripts/discourse-i18n/src/index.js#L253-L256